packages/plugin/src/browser/hosted-plugin-manager-client.ts

```diff
@@ -0,0 +1,133 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { injectable, inject } from 'inversify';
+import URI from '@theia/core/lib/common/uri';
+import { MessageService } from '@theia/core/lib/common';
+import { LabelProvider, isNative } from '@theia/core/lib/browser';
+import { WindowService } from '@theia/core/lib/browser/window/window-service';
+import { WorkspaceService } from '@theia/workspace/lib/browser';
+import { FileSystem } from '@theia/filesystem/lib/common';
+import { FileDialogFactory, DirNode } from '@theia/filesystem/lib/browser';
+import { HostedPluginServer } from '../common/plugin-protocol';
+import { HostedPluginCommands } from './plugin-api-frontend-contribution';
+
+/**
+ * Responsible for UI to set up and control Hosted Plugin Instance.
+ */
+@injectable()
+export class HostedPluginManagerClient {
+    @inject(HostedPluginServer)
+    protected readonly hostedPluginServer: HostedPluginServer;
+    @inject(MessageService)
+    protected readonly messageService: MessageService;
+    @inject(FileDialogFactory)
+    protected readonly fileDialogFactory: FileDialogFactory;
+    @inject(LabelProvider)
+    protected readonly labelProvider: LabelProvider;
+    @inject(WindowService)
+    protected readonly windowService: WindowService;
+    @inject(FileSystem)
+    protected readonly fileSystem: FileSystem;
+    @inject(WorkspaceService)
+    protected readonly workspaceService: WorkspaceService;
+
+    protected pluginLocation: URI | undefined;
+    protected pluginInstanceUri: string | undefined;
+
+    async start(): Promise<void> {
+        if (!this.pluginLocation) {
+            await this.selectPluginPath();
+            if (!this.pluginLocation) {
+                // selection was cancelled
+                return;
+            }
+        }
+        try {
+            this.messageService.info('Starting hosted instance server ...');
+            await this.doRunRequest(this.pluginLocation);
+            this.messageService.info('Hosted instanse is running at: ' + this.pluginInstanceUri);
+        } catch (error) {
+            this.messageService.error('Failed to run hosted plugin instanse: ' + this.getErrorMessage(error));
+        }
+    }
+
+    async stop(): Promise<void> {
+        try {
+            await this.hostedPluginServer.terminateHostedPluginInstance();
+            this.messageService.info((this.pluginInstanceUri ? this.pluginInstanceUri : 'The instance') + ' has been terminated.');
+        } catch (error) {
+            this.messageService.warn(this.getErrorMessage(error));
+        }
+    }
+
+    async restart(): Promise<void> {
+        if (await this.hostedPluginServer.isHostedTheiaRunning()) {
+            await this.stop();
+
+            this.messageService.info('Starting hosted instance server ...');
```

Comment:
should we display `Restarting` for this case ?

Reply:
User will see message `Stopping ...` and then `Starting...`

Reply:
not important but is it possible to only see "Restarting..." ?